File: secure_server.py
#!/usr/bin/env python3
"""
Secure HTTP server that serves static files and provides configuration API
Handles both main site content and secure config endpoints
"""

import json
import os
import mimetypes
from http.server import BaseHTTPRequestHandler, HTTPServer
from urllib.parse import urlparse

class SecureHTTPHandler(BaseHTTPRequestHandler):
    def do_GET(self):
        """Handle GET requests"""
        parsed_path = urlparse(self.path)
        
        # Handle API config endpoint
        if parsed_path.path == '/api/config':
            self.serve_config()
            return
        
        # Handle static file serving
        self.serve_static_file()
    
    def do_OPTIONS(self):
        """Handle CORS preflight requests"""
        self.send_response(200)
        self.send_header('Access-Control-Allow-Origin', '*')
        self.send_header('Access-Control-Allow-Methods', 'GET, POST, OPTIONS')
        self.send_header('Access-Control-Allow-Headers', 'Content-Type')
        self.end_headers()
    
    def serve_config(self):
        """Serve secure configuration"""
        self.send_response(200)
        self.send_header('Content-Type', 'application/json')
        self.send_header('Access-Control-Allow-Origin', '*')
        self.send_header('Cache-Control', 'no-cache')
        self.end_headers()
        
        # Public configuration that's safe to expose to client-side
        config = {
            'paypal': {
                'clientId': os.getenv('PAYPAL_CLIENT_ID', 'BAAhtnXfCO2At0RMUwWN1IzNH9YJ2iTdUB6kaInTLIuvyUXjv7WbyixHk4R7dujr_Y0AY9ZT29WKL0d6X0'),
                'environment': os.getenv('PAYPAL_ENVIRONMENT', 'sandbox')
            },
            'firebase': {
                'apiKey': os.getenv('VITE_FIREBASE_API_KEY', 'AIzaSyA78g7-5gHQ8MK4JZQjdJSZ4F2sWr0vQ_Y'),
                'authDomain': f"{os.getenv('VITE_FIREBASE_PROJECT_ID', 'resumesmartbuild')}.firebaseapp.com",
                'projectId': os.getenv('VITE_FIREBASE_PROJECT_ID', 'resumesmartbuild'),
                'storageBucket': f"{os.getenv('VITE_FIREBASE_PROJECT_ID', 'resumesmartbuild')}.appspot.com",
                'messagingSenderId': '190620294122',
                'appId': os.getenv('VITE_FIREBASE_APP_ID', '1:190620294122:web:a7b8c9d0e1f2g3h4i5j6k7')
            },
            'app': {
                'name': 'ResumeSmartBuild',
                'version': '1.0.0'
            }
        }
        
        self.wfile.write(json.dumps(config).encode('utf-8'))
    
    def serve_static_file(self):
        """Serve static files"""
        # Remove query parameters
        path = self.path.split('?')[0]
        
        # Default to index.html for root
        if path == '/':
            path = '/index.html'
        
        # Remove leading slash for file system
        file_path = path.lstrip('/')
        
        try:
            # Check if file exists
            if os.path.isfile(file_path):
                # Get MIME type
                content_type, _ = mimetypes.guess_type(file_path)
                if content_type is None:
                    content_type = 'text/plain'
                
                # Read and serve file
                with open(file_path, 'rb') as f:
                    content = f.read()
                
                self.send_response(200)
                self.send_header('Content-Type', content_type)
                self.send_header('Content-Length', str(len(content)))
                
                # Add cache headers for static assets
                if file_path.endswith(('.css', '.js', '.png', '.jpg', '.jpeg', '.gif', '.ico')):
                    self.send_header('Cache-Control', 'public, max-age=3600')
                
                self.end_headers()
                self.wfile.write(content)
            else:
                # File not found
                self.send_response(404)
                self.send_header('Content-Type', 'text/html')
                self.end_headers()
                self.wfile.write(b'<h1>404 - File Not Found</h1>')
                
        except Exception as e:
            # Server error
            self.send_response(500)
            self.send_header('Content-Type', 'text/html')
            self.end_headers()
            self.wfile.write(f'<h1>500 - Server Error</h1><p>{str(e)}</p>'.encode('utf-8'))

if __name__ == '__main__':
    port = int(os.getenv('PORT', 5000))
    server = HTTPServer(('0.0.0.0', port), SecureHTTPHandler)
    print(f'Secure server running on port {port}')
    print('Endpoints:')
    print('  GET  / - Static file serving')
    print('  GET  /api/config - Secure configuration')
    server.serve_forever()

File: secure_server.js
// secure_server.js - Full Node.js Express App for ResumeSmartBuild
const express = require('express');
const cors = require('cors');
const bodyParser = require('body-parser');
const sgMail = require('@sendgrid/mail');
const path = require('path');

const app = express();
const PORT = process.env.PORT || 5000;

// Configure SendGrid
sgMail.setApiKey(process.env.SENDGRID_API_KEY);

// Middleware
app.use(cors({
  origin: '*',
  methods: ['GET', 'POST', 'PUT', 'DELETE', 'OPTIONS'],
  allowedHeaders: ['Content-Type', 'Authorization']
}));
app.use(bodyParser.json());
app.use(bodyParser.urlencoded({ extended: true }));

// Serve static files from public directory
app.use(express.static('public'));

// Additional static file serving for current structure
app.use('/styles', express.static('styles'));
app.use('/scripts', express.static('scripts'));
app.use('/templates', express.static('templates'));
app.use('/articles', express.static('articles'));

// Health check endpoint
app.get('/health', (req, res) => {
  res.status(200).send('OK');
});

// Main route - serve index.html
app.get('/', (req, res) => {
  const indexPath = path.join(__dirname, 'public', 'index.html');
  
  // Check if public/index.html exists, otherwise serve root index.html
  const fs = require('fs');
  if (fs.existsSync(indexPath)) {
    res.sendFile(indexPath);
  } else {
    res.sendFile(path.join(__dirname, 'index.html'));
  }
});

// Newsletter signup endpoint
app.post('/api/newsletter', async (req, res) => {
  const { email, name } = req.body;

  if (!email) {
    return res.status(400).json({ 
      success: false, 
      message: 'Email is required' 
    });
  }

  try {
    const msg = {
      to: email,
      from: 'info@resumesmartbuild.com',
      subject: 'Welcome to ResumeSmartBuild!',
      html: `
        <div style="font-family: Arial, sans-serif; max-width: 600px; margin: 0 auto;">
          <h1 style="color: #2c3e50;">Welcome to ResumeSmartBuild!</h1>
          <p>Hi ${name || 'there'},</p>
          <p>Thanks for signing up for our newsletter! We're excited to help you build smarter resumes with AI-powered tools.</p>
          
          <div style="background: #f8f9fa; padding: 20px; border-radius: 8px; margin: 20px 0;">
            <h3>🚀 What's Next?</h3>
            <ul>
              <li>Try our ATS Resume Scanner</li>
              <li>Browse premium resume templates</li>
              <li>Use our AI-powered job matching</li>
              <li>Access our resume optimization tools</li>
            </ul>
          </div>
          
          <p><a href="https://resumesmartbuild.com" style="background: #007bff; color: white; padding: 12px 24px; text-decoration: none; border-radius: 5px;">Get Started Now</a></p>
          
          <p>Best regards,<br>The ResumeSmartBuild Team</p>
        </div>
      `,
    };

    await sgMail.send(msg);
    console.log(`✅ Newsletter email sent to: ${email}`);
    
    res.status(200).json({ 
      success: true, 
      message: 'Thank you! Check your inbox for a welcome email.' 
    });
  } catch (error) {
    console.error('❌ SendGrid error:', error.response?.body || error.message);
    res.status(500).json({ 
      success: false, 
      message: 'Failed to send email. Please try again later.' 
    });
  }
});

// API Config endpoint for Firebase/PayPal configuration
app.get('/api/config', (req, res) => {
  res.json({
    paypal: {
      clientId: process.env.PAYPAL_CLIENT_ID || 'BAAhtnXfCO2At0RMUwWN1IzNH9YJ2iTdUB6kaInTLIuvyUXjv7WbyixHk4R7dujr_Y0AY9ZT29WKL0d6X0',
      environment: process.env.PAYPAL_ENVIRONMENT || 'sandbox'
    },
    firebase: {
      apiKey: process.env.VITE_FIREBASE_API_KEY,
      authDomain: `${process.env.VITE_FIREBASE_PROJECT_ID || 'resumesmartbuild'}.firebaseapp.com`,
      projectId: process.env.VITE_FIREBASE_PROJECT_ID || 'resumesmartbuild',
      storageBucket: `${process.env.VITE_FIREBASE_PROJECT_ID || 'resumesmartbuild'}.appspot.com`,
      messagingSenderId: '190620294122',
      appId: process.env.VITE_FIREBASE_APP_ID || '1:190620294122:web:9a93a5763ddcf3e1c63093'
    },
    app: {
      name: 'ResumeSmartBuild',
      version: '1.0.0'
    }
  });
});

// Future AI Assistant endpoint placeholder
app.post('/api/ai/query', (req, res) => {
  const { message, context } = req.body;
  
  // Placeholder for AI integration (Chatbase, OpenAI, etc.)
  res.json({
    success: true,
    message: 'AI assistant integration coming soon!',
    response: 'This endpoint will be connected to AI services soon.',
    context: context || 'general'
  });
});

// Future Admin Dashboard endpoints
app.get('/api/admin/analytics', (req, res) => {
  // Placeholder for admin analytics
  res.json({
    totalUsers: 0,
    newsletterSignups: 0,
    atsScans: 0,
    premiumSubscriptions: 0,
    message: 'Admin dashboard coming soon'
  });
});

app.get('/api/admin/entries', (req, res) => {
  // Placeholder for form entries
  res.json({
    entries: [],
    total: 0,
    message: 'Form entries tracking coming soon'
  });
});

app.get('/api/admin/logs', (req, res) => {
  // Placeholder for error logs
  res.json({
    logs: [],
    errors: 0,
    message: 'Error logging dashboard coming soon'
  });
});

// Catch-all route for SPA
app.get('*', (req, res) => {
  const publicIndexPath = path.join(__dirname, 'public', 'index.html');
  const rootIndexPath = path.join(__dirname, 'index.html');
  
  const fs = require('fs');
  if (fs.existsSync(publicIndexPath)) {
    res.sendFile(publicIndexPath);
  } else if (fs.existsSync(rootIndexPath)) {
    res.sendFile(rootIndexPath);
  } else {
    res.status(404).send('Page not found');
  }
});

// Error handling middleware
app.use((err, req, res, next) => {
  console.error('❌ Server error:', err.stack);
  res.status(500).json({
    success: false,
    message: 'Internal server error',
    error: process.env.NODE_ENV === 'development' ? err.message : 'Something went wrong'
  });
});

// Start server
app.listen(PORT, '0.0.0.0', () => {
  console.log(`🚀 ResumeSmartBuild server running on port ${PORT}`);
  console.log(`📧 SendGrid configured: ${process.env.SENDGRID_API_KEY ? '✅' : '❌'}`);
  console.log(`🌐 Access at: http://localhost:${PORT}`);
});